include/bluetooth/services/ots.h

```diff
@@ -0,0 +1,600 @@
+ */
+#define BT_GATT_OTS_OLCP_SET_FEAT_GO_TO(feat) \
+	WRITE_BIT(feat, BT_GATT_OTS_OLCP_FEAT_GO_TO, 1)
+
+/** @brief Set @ref BT_GATT_OTS_OLCP_FEAT_ORDER feature.
+ *
+ *  @param feat OTS features
+ */
+#define BT_GATT_OTS_OLCP_SET_FEAT_ORDER(feat) \
+	WRITE_BIT(feat, BT_GATT_OTS_OLCP_FEAT_ORDER, 1)
+
+/** @brief Set @ref BT_GATT_OTS_OLCP_FEAT_NUM_REQ feature.
+ *
+ *  @param feat OTS features
+ */
+#define BT_GATT_OTS_OLCP_SET_FEAT_NUM_REQ(feat) \
+	WRITE_BIT(feat, BT_GATT_OTS_OLCP_FEAT_NUM_REQ, 1)
+
+/** @brief Set @ref BT_GATT_OTS_OLCP_FEAT_CLEAR feature.
+ *
+ *  @param feat OTS features
+ */
+#define BT_GATT_OTS_OLCP_SET_FEAT_CLEAR(feat) \
+	WRITE_BIT(feat, BT_GATT_OTS_OLCP_FEAT_CLEAR, 1)
+
+/** @brief Get @ref BT_GATT_OTS_OLCP_GET_FEAT_GO_TO feature.
+ *
+ *  @param feat OTS features
+ */
+#define BT_GATT_OTS_OLCP_GET_FEAT_GO_TO(feat) \
+	((feat) & BIT(BT_GATT_OTS_OLCP_FEAT_GO_TO))
+
+/** @brief Get @ref BT_GATT_OTS_OLCP_GET_FEAT_ORDER feature.
+ *
+ *  @param feat OTS features
+ */
+#define BT_GATT_OTS_OLCP_GET_FEAT_ORDER(feat) \
+	((feat) & BIT(BT_GATT_OTS_OLCP_FEAT_ORDER))
+
+/** @brief Get @ref BT_GATT_OTS_OLCP_GET_FEAT_NUM_REQ feature.
+ *
+ *  @param feat OTS features
+ */
+#define BT_GATT_OTS_OLCP_GET_FEAT_NUM_REQ(feat) \
+	((feat) & BIT(BT_GATT_OTS_OLCP_FEAT_NUM_REQ))
+
+/** @brief Get @ref BT_GATT_OTS_OLCP_GET_FEAT_CLEAR feature.
+ *
+ *  @param feat OTS features
+ */
+#define BT_GATT_OTS_OLCP_GET_FEAT_CLEAR(feat) \
+	((feat) & BIT(BT_GATT_OTS_OLCP_FEAT_CLEAR))
+
+/**@brief Features of the OTS. */
+struct bt_gatt_ots_feat {
+	/* OACP Features */
+	uint32_t oacp;
+
+	/* OLCP Features */
+	uint32_t olcp;
+};
+
+/** @brief Opaque OTS instance. */
+struct bt_gatt_ots;
+
+/** @brief OTS server callback structure. */
+struct bt_gatt_ots_cb {
+	/** Object created callback
+	 *
+	 *  This callback is called whenever a new object is created.
+	 *  Application can reject this request by returning an error
+	 *  when it does not have necessary resources to hold this new
+	 *  object.
+	 *
+	 *  @param ots  OTS Server instance
+	 *  @param conn The connection that is requesting object creation or
+	 *              NULL if object is created by bt_gatt_ots_object_add()
+	 *  @param id   Object ID
+	 *  @param init Object initialization metadata
+	 *
+	 *  @return 0 in case of success or negative value in case of error.
+	 *  Possible return values:
+	 *  -ENOMEM if no available space for new object.
+	 */
+	int (*obj_created)(struct bt_gatt_ots *ots, struct bt_conn *conn,
+			   uint64_t id,
+			   const struct bt_gatt_ots_obj_init *init);
+
+	/** Object deleted callback
+	 *
+	 *  This callback is called whenever an object is deleted.
+	 *
+	 *  @param ots  OTS Server instance
+	 *  @param conn The connection that deleted the object or NULL if
+	 *              this request came from the server
+	 *  @param id   Object ID
+	 */
+	void (*obj_deleted)(struct bt_gatt_ots *ots, struct bt_conn *conn,
+			    uint64_t id);
```

Comment:
Should the application be able to refuse a deletion?

Reply:
It is possible to make an undeletable object by setting its properties correctly. If we added an API to change the object properties dynamically then it wouldn't be necessary to change this callback. Does it seem to you like a good solution?

Reply:
Right, if the application wants to make an object undeletable, it can just set the properties of `bt_gatt_ots_obj_init` to that. Good point :) No need to update the API, unless there is a usecase for the application to perhaps only the device that created the object, delete it (e.g. by validating against the `conn`). 

Can imagine such a scenario where:
Client A creates ObjA, but where Client B is not allowed, by the application, to delete ObjA, but Client A is allowed to? 

If yes, then I think the API should allow for the application to deny a deletion, else it's good as it is.

Reply:
Let's keep it simple, OTS specification does not mention individual client permissions like superuser and standard one. I will tag this API as experimental so that we can change it in the future